Automate hash scans

bra1ncramp opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

Create extra file to automate the IOC Hash scans.

* This script currently is only valid for Debian and Fedora systems.
* Requires AWS_CREDENTIALS_FILE and AWS_REGION environmental variables to already be set.
* Assumes it will be run from `ioc-scanner/extras`
* Takes a file with a list of instance id's in separate lines
* Requires declaring the AWS_PROFILE as an argument
* Starts port forwarding on the local system. Then it verifies that `netcat` is installed in the instance and starts `netcat` listening on port `6666`. It uploads the latest copy of `ioc_scanner.py` with the latest hashes to the instance. Then it executes `ioc_scanner.py` and directs the output to a local log file.

## 💭 Motivation and context ##

The current process requires manually uploading the `ioc_scanner.py` file to each instance one at a time. This is a laborious method and not scaleable.

## 🧪 Testing ##

This was tested using the default test blobs in `ioc_scanner.py`.

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

bra1ncramp committed:
Automate hash scans

bra1ncramp committed:
Uncomment actual IOC scan

This part was commented out durring development testing. Accidentally left that way at commit.

bra1ncramp committed:
Trim trailing whitespace

bra1ncramp committed:
Fix comments

bra1ncramp committed:
Convert spaced to tabs to pass shfmt check

bra1ncramp committed:
Fix spacing to pass pre-commit

bra1ncramp committed:
Merge branch 'develop' into automate-hash-scans

dav3r commented:
For the [same reasons as discussed](https://github.com/cisagov/ioc-scanner/pull/53#issuecomment-1789452452) in #53, this PR will not require a version change to this repo, so I deleted the "Pre-merge" and "Post-merge" checklist sections in the PR description.

jsf9k commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+#!/bin/bash
+#
+##Add the hashes to the blob at src/ioc_scan/ioc_scanner.py
+#
+#
+#  The filename specified in the first argument
```
```suggestion
# Add the hashes to the blob at src/ioc_scan/ioc_scanner.py
#
#
# The filename specified in the first argument
```

jsf9k commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+logfile="$HOME/$today-ioc-scanner-hashscan.log"
+
+# Suppress some verbose stdout.
+# Suppress stnderr of pkill command
```
Here's a suggestion to fix a typo.
```suggestion
# Suppress stderr of pkill command
```

Is it wise to throw away output like this?  What if something goes wrong when the script is running?

bra1ncramp replied:
That's the rub here. I don't want to through it away - I'd prefer to temporarily suppress during the `pkill` command and then restart it. But all the methods I can find don't work. Probably because the command runs in a sub-shell - though I can't be certain of the exact reason.

jsf9k replied:
Please see my suggestion [here](https://github.com/cisagov/ioc-scanner/pull/54#pullrequestreview-1730755677).

jsf9k replied:
Looks like the typo has been fixed.

jsf9k commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+      aws --profile="$AWSPROF" --region="$AWS_REGION" \
+      ssm start-session --target="$i" \
+      --document=AWS-StartInteractiveCommand \
+      --parameters="command='sudo apt-get install netcat -y'"
```
```suggestion
      --parameters="command='sudo apt-get --yes install netcat'"
```

jsf9k commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+      aws --profile="$AWSPROF" --region="$AWS_REGION" \
+      ssm start-session --target="$i" \
+      --document=AWS-StartInteractiveCommand \
+      --parameters="command='sudo yum install netcat -y'"
```
Modern RH-based systems use `dnf`, not `yum`.
```suggestion
      --parameters="command='sudo dnf --assumeyes install netcat'"
```

jsf9k commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+      aws --profile="$AWSPROF" --region="$AWS_REGION" \
+      ssm start-session --target="$i" \
+      --document=AWS-StartInteractiveCommand \
+      --parameters="command='cd ~/src/ioc_scan; nc -l -p 6666 | tar xzf -'"
```
```suggestion
      --parameters="command='cd ~/src/ioc_scan; nc --listen --local-port 6666 | tar --extract --gzip --file -'"
```

bra1ncramp replied:
I'm not so sure about this. Debian doesn't have those long form options for `nc` (at least according to their man page).

jsf9k replied:
Please see my suggestion [here](https://github.com/cisagov/ioc-scanner/pull/54#pullrequestreview-1730755677).

bra1ncramp replied:
See 5deb667927aafd66d09c846733ecd78979c6ddeb

jsf9k commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+      aws --profile="$AWSPROF" --region="$AWS_REGION" \
+      ssm start-session --target="$i" \
+      --document=AWS-StartInteractiveCommand \
+      --parameters="command='cd ~/src/ioc_scan; nc -l 6666 | tar xzf -'"
```
```suggestion
      --parameters="command='cd ~/src/ioc_scan; nc --listen 6666 | tar --extract --gzip --file -'"
```

bra1ncramp replied:
Fedora doesn't have long form options for `nc` (at least according to it's man page).

jsf9k replied:
Please see my suggestion [here](https://github.com/cisagov/ioc-scanner/pull/54#pullrequestreview-1730755677).

bra1ncramp replied:
see 05c6a7fd700355a0643e1e46912dc2673c7b8be2

jsf9k commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+    aws --profile="$AWSPROF" --region="$AWS_REGION" \
+    ssm start-session --target="$i" \
+    --document=AWS-StartInteractiveCommand \
+    --parameters="command='if [ ! -d ~/src/ioc_scan ]; then mkdir -p ~/src/ioc_scan; fi'"
```
```suggestion
    --parameters="command='if [ ! -d ~/src/ioc_scan ]; then mkdir --parents ~/src/ioc_scan; fi'"
```

jsf9k commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+  cd "$pydir" || exit
+
+  echo "Upload lastest ioc_scanner.py to $instanceName"
+  tar czf - ./ioc_scanner.py | nc localhost 5555
```
```suggestion
  tar --create --gzip --file - ./ioc_scanner.py | nc localhost 5555
```

jsf9k commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+    --parameters="command=python3 ~/src/ioc_scan/ioc_scanner.py" >> "$logfile"
+
+  # Killing port forwading so we can do this again on the next Instance.
+  while pgrep -fq session-manager-plugin; do
```
```suggestion
  while pgrep --full --queue session-manager-plugin; do
```

bra1ncramp replied:
1. macOS does not have those long form options
2. `-q` is for `quiet`, not `queue`

jsf9k replied:
Doesn't this run on Debian or Fedora, though, not MacOS?

jsf9k replied:
According to `man pgrep` on my local system, `-q` is identical to `--queue`.  Interesting that it is different in Debian and Fedora.

jsf9k replied:
I see that this piece runs on OSX.

jsf9k commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+
+##clean up log output for readability
+while grep --quiet --ignore-case "session" "$logfile"; do
+  sed -i '' '/session/d' "$logfile"
```
```suggestion
  sed --in-place '' '/session/d' "$logfile"
```

bra1ncramp replied:
macOS does not have long-form options for `sed`

jsf9k replied:
Doesn't this run on Debian or Fedora, though, not MacOS?

jsf9k replied:
Nevermind.

jsf9k commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+  sed -i '' '/session/d' "$logfile"
+done
+
+sed -i '' 'N;/^\n$/D;P;D;' "$logfile"
```
```suggestion
sed --in-place '' 'N;/^\n$/D;P;D;' "$logfile"
```

bra1ncramp replied:
macOS does not have long-form options for `sed`

jsf9k replied:
Doesn't this run on Debian or Fedora, though, not MacOS?

jsf9k replied:
Nevermind.

jsf9k commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+    --document=AWS-StartInteractiveCommand \
+    --parameters="command=cat /etc/os-release | grep --extended-regexp ^NAME=" \
+    --output text)
+  echo "$OSNAME" | grep NAME | cut -d'"' -f2 | cut -d' ' -f1
```
```suggestion
  echo "$OSNAME" | grep NAME | cut --delimiter='"' --fields=2 | cut --delimiter=' ' --fields=1
```

bra1ncramp replied:
macOS does not have long-form options for `cut`

jsf9k replied:
Doesn't this run on Debian or Fedora, though, not MacOS?

jsf9k replied:
Ah, I see.  This part runs on your local machine.

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+#  The filename specified in the first argument
+# (instance-list-file) should contain a list of instance id strings, one per line.
+#
+# Usage: ./ioc_hash_scan.sh instance-list-file <AWS_PROFILE>
```
I would move the usage text to just after the shebang line.  Also, it's customary to put a blurb before the usage explaining what the script actually does, like we do here: https://github.com/cisagov/ioc-scanner/blob/1eac84e990aa94230ef14c335bc736fbfc1c5aa4/extras/ioc_scan_by_host.sh#L1-L7

bra1ncramp replied:
Well, that's basically what I've done. I can move the comment about adding hashes to the blob to after the usage, and be more verbose about the purpose. But I literally ripped this header from the example you sited above.

I'll work on making it cleaner.

dav3r replied:
My main point is that the comment blurb does not adequately explain what this script does.  Let me know when you've updated it and I'll take a look.

bra1ncramp replied:
Check 153ca1c9fe3444da32c8ea1d33b7b6113579c14f

dav3r replied:
Much better, thanks!

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+#!/bin/bash
+#
+##Add the hashes to the blob at src/ioc_scan/ioc_scanner.py
```
Can this line be deleted?  If not, it should be moved somewhere more relevant and conform to our standard comment style:
```python
# Add the hashes to the blob at src/ioc_scan/ioc_scanner.py
```

bra1ncramp replied:
I will find a better place to move it to.

bra1ncramp replied:
check 153ca1c9fe3444da32c8ea1d33b7b6113579c14f

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+#  The filename specified in the first argument
+# (instance-list-file) should contain a list of instance id strings, one per line.
+#
+# Usage: ./ioc_hash_scan.sh instance-list-file <AWS_PROFILE>
```
Is there a reason why the instance IDs cannot be passed in as arguments to this script?

Also, is there a reason why you are not just getting `AWS_PROFILE` as an environment variable, similar to `AWS_CREDENTIALS_FILE` and `AWS_REGION`?

bra1ncramp replied:
I thought about grabbing `AWS_PROFILE` as an environmental profile - but I'm not sure how to do that well. Consider that there are likely a dozen or so possible `AWS_PROFILE`'s that someone may have - setting a single one as an environmental profile could create unnecessary pain and suffering.

I get this is likely the roughest part. I would like to simplify it.

dav3r replied:
> Consider that there are likely a dozen or so possible `AWS_PROFILE`'s that someone may have - setting a single one as an environmental profile could create unnecessary pain and suffering.

It's trivial to call the script with only the profile and other env variables that you want, like so:
```console
AWS_SHARED_CREDENTIALS_FILE=~/.aws/my-creds AWS_REGION=us-east-1 AWS_PROFILE=my-profile ./ioc_hash_scan.sh i-0123456789abcdef
```

bra1ncramp replied:
See a2162cf552655c652e149bbd1611652eb3401170

dav3r replied:
Typically, you don't include environment variables in example usage (e.g. [here](https://github.com/cisagov/ioc-scanner/blob/a2162cf552655c652e149bbd1611652eb3401170/extras/ioc_hash_scan.sh#L8) and [here](https://github.com/cisagov/ioc-scanner/blob/a2162cf552655c652e149bbd1611652eb3401170/extras/ioc_hash_scan.sh#L22)), so I suggest removing them from those lines.

Instead, I think you should mention in the [script's description comment](https://github.com/cisagov/ioc-scanner/blob/a2162cf552655c652e149bbd1611652eb3401170/extras/ioc_hash_scan.sh#L3-L6) that it expects certain environment variables to be set and be clear about which environment variables are needed.

Additionally, since those environment variables are required for this script to function properly, the script should check that they are indeed set and throw an error if they are not.

bra1ncramp replied:
I'm not sure how to test for the variables correctly. 
To test, I added this to the script, but did not pass the variable
`[[ -z "$AWS_PROFILE" ]] && { echo "AWS_PROFILE environmental variable is not set. Exiting"; exit 1; }`
But instead of the error output I expected, I got
`AWS_PROFILE: unbound variable`

bra1ncramp replied:
See 06d01ea7e1e3643be6681d7e562cdc63d08ac083

dav3r replied:
Much better, thanks!

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+#
+# Usage: ./ioc_hash_scan.sh instance-list-file <AWS_PROFILE>
+#
+# Must have AWS_CREDENTIAL_FILE and AWS_REGION exported to environmental varialble
```
Correct environment variable name, reword slightly, and fix typo:
```suggestion
# Must have AWS_CREDENTIALS_FILE and AWS_REGION available as environmental variables
```

bra1ncramp replied:
There could be multiple `AWS_CREDENTIALS_FILE`'s - I used the different variable name intentionally. I could include a comment here, though.

dav3r replied:
We don't want to deal with multiple credentials files.  I would just use the standard AWS env. variable (`AWS_SHARED_CREDENTIALS_FILE`).

bra1ncramp replied:
see c5dbad1a8a880c57d53c04202ece97d2dd6fb576 and 8c75300af2c99afdd9a2ed6523638a8999559482

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+set -o errexit
+set -o pipefail
+
+# Directory path of the ioc_scanner.py
```
Improve comment wording:
```suggestion
# Directory path of the ioc_scanner.py file
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+# source ~/.bashrc
+#
+# This script assumes that it exists in the ioc-scanner/extras/ directory.
+# If this is different, please edit the variable $pydir to point to
```
Improve verbiage:
```suggestion
# If it does not, please edit the variable $pydir to point to
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+done < "$1"
+
+if [ ${#serverList[@]} -eq 0 ]; then
+  echo No Instances found in "$1" - exiting.
```
Remove unneeded capitalization:
```suggestion
  echo No instances found in "$1" - exiting.
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+  exit 1
+fi
+
+# Path of aws credentials file
```
Hopefully this goes away (see my previous comment), but if not, this comment should be corrected:
```suggestion
# Get AWS profile from command line argument
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+  HOST=$(AWS_SHARED_CREDENTIALS_FILE="$AWS_CREDENTIALS_FILE" \
+    aws --profile="$AWSPROF" --region="$AWS_REGION" \
+    ssm start-session --target="$i" \
+    --document=AWS-StartInteractiveCommand \
```
Simply running `hostname` should work fine with `AWS-StartNonInteractiveCommand`:
```suggestion
    --document=AWS-StartNonInteractiveCommand \
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+for i in "${serverList[@]}"; do
+  OS="$(getOSType)"
+  if [[ "$OS" != "Debian" && "$OS" != "Fedora" ]]; then
+    echo "Non-supported OS Type"
```
Improve error message:
```suggestion
    echo "Instance $i is running an operating system ($OS) that is not supported by this script."
    echo "Currently only Debian and Fedora are supported - exiting."
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+
+  instanceName=$(getHost)
+
+  echo "Begining scan of Instance $i -- $instanceName" | tee -a "$logfile"
```
Fix typo, remove unnecessary capitalization, use long name for flag:
```suggestion
  echo "Beginning scan of instance $i -- $instanceName" | tee --append "$logfile"
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+  echo "Begining scan of Instance $i -- $instanceName" | tee -a "$logfile"
+
+  # Start Port Forwarding
+  echo "Begining port forwarding"
```
Fix typo and add more info:
```suggestion
  echo "Beginning port forwarding (local port 5555 to remote port 6666)"
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+  echo "Begining port forwarding"
+  portForward &
+
+  # Create ~/src/ioc_scan directory on Instance
```
Remove unnecessary capitalization:
```suggestion
  # Create ~/src/ioc_scan directory on instance
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+    --document=AWS-StartInteractiveCommand \
+    --parameters="command='if [ ! -d ~/src/ioc_scan ]; then mkdir -p ~/src/ioc_scan; fi'"
+
+  #Install netcat and start listening on port 6666
```
Use our standard comment spacing:
```suggestion
  # Install netcat and start listening on port 6666
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+  #Install netcat and start listening on port 6666
+  echo "Verifying netcat on $instanceName"
+  installNC
+  echo "Begin listening on $instanceName"
```
Add port number to output message:
```suggestion
  echo "Begin listening on $instanceName (port 6666)"
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+  echo "Begin listening on $instanceName"
+  startListen &
+
+  # Copy latest ioc_scanner.py to target instance
+  curdir=$(pwd)
+  cd "$pydir" || exit
+
+  echo "Upload lastest ioc_scanner.py to $instanceName"
```
Remove an unneeded word from a comment and an output message:
```suggestion
  # Copy ioc_scanner.py to target instance
  curdir=$(pwd)
  cd "$pydir" || exit

  echo "Upload ioc_scanner.py to $instanceName"
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+  cd "$curdir" || exit
+
+  # Run ioc_scanner.py on target instance
+  echo "Scan $instanceName for IOC Hashes"
```
Remove unnecessary capitalization:
```suggestion
  echo "Scan $instanceName for IOC hashes"
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+    --document=AWS-StartInteractiveCommand \
+    --parameters="command=python3 ~/src/ioc_scan/ioc_scanner.py" >> "$logfile"
+
+  # Killing port forwading so we can do this again on the next Instance.
```
Fix typo and remove unnecessary capitalization:
```suggestion
  # Killing port forwarding so we can do this again on the next instance.
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+  # Killing port forwading so we can do this again on the next Instance.
+  while pgrep -fq session-manager-plugin; do
+    pkill session-manager-plugin
+    # We need to wait, as some race conditions can occure.
```
Fix typo:
```suggestion
    # We need to wait, as some race conditions can occur.
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+  echo "------------------------------------------------------------------------" | tee -a "$logfile"
+done
+
+##clean up log output for readability
```
Use our standard comment spacing:
```suggestion
# Clean up log output for readability
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+    # We need to wait, as some race conditions can occure.
+    sleep 5
+  done
+  echo "------------------------------------------------------------------------" | tee -a "$logfile"
```
```suggestion
  echo "------------------------------------------------------------------------" | tee --append "$logfile"
```

bra1ncramp replied:
macOS does not have long-form options for `tee`

dav3r replied:
Dangit!  Retracted. 👍

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+    --parameters="command=python3 ~/src/ioc_scan/ioc_scanner.py" >> "$logfile"
+
+  # Killing port forwading so we can do this again on the next Instance.
+  while pgrep -fq session-manager-plugin; do
```
Did you try using [`aws ssm terminate-session`](https://docs.aws.amazon.com/cli/latest/reference/ssm/terminate-session.html) to stop the port-forwarding session?  That seems cleaner than killing the `session-manager-plugin`.

bra1ncramp replied:
I did not know about that option. I will check it out. That could save me a headache!

bra1ncramp replied:
`TerminateSession` is not allowed by the `StartStopSSMSession` role

jsf9k replied:
@dav3r - This seems like a permission we should add to `StartStopSSMSession`.  What do you think?

dav3r replied:
I was just looking at this and as it turns out, [we already do include it](https://github.com/cisagov/session-manager-tf-module/blob/dc73f30712dfb8609346c396202875c5d085331d/ssm_session_policy.tf#L58-L66):
```
  # Allow the user to terminate his or her own sessions
  statement {
    actions = [
      "ssm:TerminateSession",
    ]
    resources = [
      "arn:aws:ssm:${var.aws_region}:${local.this_account_id}:session/${aws:username}-*",
    ]
  }
```

I'm trying to figure out why it wouldn't be working as expected here.

dav3r replied:
Aha, here's the issue.  Since we are using an assumed role, the resource in question looks like this:
```
arn:aws:sts::${local.this_account_id}:assumed-role/StartStopSSMSession/&{aws:username}-*
```
and not like this (as currently defined in the policy mentioned above):
```
arn:aws:ssm:${var.aws_region}:${local.this_account_id}:session/&{aws:username}-*
```

dav3r replied:
For additional troubleshooting/brainstorming, here is the session that @bra1ncramp attempted to terminate and failed:
```
arn:aws:ssm:us-east-1:<acct-number-redacted>:session/<username-redacted>-0fac6447f6dfdabf9 
```

Here is the full (sanitized) error message that @bra1ncramp sent to me:
```
An error occurred (AccessDeniedException) when calling the TerminateSession operation: 
User: arn:aws:sts::<acct-number-redacted>:assumed-role/StartStopSSMSession/<username-redacted>
is not authorized to perform: ssm:TerminateSession on resource: 
arn:aws:ssm:us-east-1:<acct-number-redacted>:session/<username-redacted>-0fac6447f6dfdabf9
because no identity-based policy allows the ssm:TerminateSession action
```

dav3r replied:
Further sanitized info from CloudTrail that might be useful for us here:
```
{
    "eventVersion": "1.08",
    "userIdentity": {
        "type": "AssumedRole",
        "principalId": "<id-redacted>:<username-redacted>",
        "arn": "arn:aws:sts::<acct-number-redacted>:assumed-role/StartStopSSMSession/<username-redacted>",
        "accountId": "<acct-number-redacted>",
        "accessKeyId": "<redacted>",
        "sessionContext": {
            "sessionIssuer": {
                "type": "Role",
                "principalId": "<id-redacted>",
                "arn": "arn:aws:iam::<acct-number-redacted>:role/StartStopSSMSession",
                "accountId": "<redacted>",
                "userName": "StartStopSSMSession"
            },
            "attributes": {
                "creationDate": "2023-11-15T20:16:11Z",
                "mfaAuthenticated": "false"
            }
        }
    },
    "eventTime": "2023-11-15T20:43:53Z",
    "eventSource": "ssm.amazonaws.com",
    "eventName": "TerminateSession",
    "awsRegion": "us-east-1",
    "sourceIPAddress": "<redacted>",
    "userAgent": "aws-cli/2.11.19 Python/3.11.3 Darwin/23.0.0 exe/x86_64 prompt/off command/ssm.terminate-session",
    "errorCode": "AccessDenied",
    "errorMessage": "User: arn:aws:sts::<acct-number-redacted>:assumed-role/StartStopSSMSession/<username-redacted> is not authorized to perform: ssm:TerminateSession on resource: arn:aws:ssm:us-east-1:<acct-number-redacted>:session/<username-redacted>-02a342dc27751ab5c because no identity-based policy allows the ssm:TerminateSession action",
    "requestParameters": null,
    "responseElements": null,
    "requestID": "<redacted>",
    "eventID": "<redacted>",
    "readOnly": false,
    "eventType": "AwsApiCall",
    "managementEvent": true,
    "recipientAccountId": "<redacted>",
    "eventCategory": "Management",
    "tlsDetails": {
        "tlsVersion": "TLSv1.2",
        "cipherSuite": "ECDHE-RSA-AES128-GCM-SHA256",
        "clientProvidedHostHeader": "ssm.us-east-1.amazonaws.com"
    }
}
```

I think the issue here is that the `${aws:username}` variable in our `StartStopSSMSession` policy is not expanding to the username of the IAM user who assumes our SSM role, and since they don't match, the `TerminateSession` action is not being allowed.  That being said, I'm not really sure how to fix this.  

@jsf9k - any thoughts?

jsf9k replied:
@bra1ncramp - Where are you testing this?  Can you try changing
```hcl
  # Allow the user to terminate his or her own sessions
  statement {
    actions = [
      "ssm:TerminateSession",
    ]
    resources = [
      "arn:aws:ssm:${var.aws_region}:${local.this_account_id}:session/&{aws:username}-*",
    ]
  }
```
to
```hcl
  # Allow the user to terminate his or her own sessions
  statement {
    actions = [
      "ssm:TerminateSession",
    ]
    resources = [
      "arn:aws:ssm:${var.aws_region}:${local.this_account_id}:session/${aws:username}-*",
    ]
  }
```
in the policy `StartStopSSMSession`?  Note that the only change is that the ampersand (`&`) is changed to a dollar sign (`$`) toward the end of the resource ARN; I provided some additional code for context.

You can simply make this change live, in the AWS console for now.

jsf9k replied:
Nevermind, @bra1ncramp.  I was wrong.

jsf9k replied:
I created cisagov/session-manager-tf-module#29, which should resolve the issue and allow users to terminate their own sessions.  @braincramp - you can apply these changes manually by editing the policy document in your CEIL account if you want to test.  The changes are already applied to the COOL staging Shared Services account, if you want to test there instead.

bra1ncramp replied:
Pending merge of https://github.com/cisagov/session-manager-tf-module/pull/29 before I can test.

jsf9k replied:
cisagov/session-manager-tf-module#29 has been merged and the changes have been applied to COOL Shared Services accounts (both production and staging).  @bra1ncramp - You should be able to test now, either in COOL staging or by applying the changes to the CEIL.

dav3r replied:
Once this last item gets resolved, I think this PR should be done.  @bra1ncramp please let us know when you've tested this (or if you have any blockers to your testing that we can help with).

bra1ncramp replied:
You may have been waiting on me for this. I have tested it and it is working. I was able to use it successfully, yesterday.

dav3r replied:
Huzzah - glorious news! 🎉

dav3r reviewed (changes requested):
Thanks for building out this script- it will be an excellent improvement from our current manual process.

Please take a look at my first round of suggestions.

dav3r commented:
@bra1ncramp Just pinging on this PR so that it doesn't get forgotten.  It's good stuff- we should try to get it across the finish line soon.

bra1ncramp committed:
Clean up comment

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

bra1ncramp committed:
Correct spelling

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

bra1ncramp committed:
use long-form options in command

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

bra1ncramp committed:
Use `dnf` and long form options

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

bra1ncramp committed:
Use long-form options in command

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

bra1ncramp committed:
Use long-form options in command

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

bra1ncramp committed:
Improve comment wording

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Improve verbiage

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

jsf9k commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,189 @@
+    --document=AWS-StartPortForwardingSession --parameters="localPortNumber=5555,portNumber=6666"
+}
+
+function installNC() {
+  if [[ "$OS" == "Debian" ]]; then
+    AWS_SHARED_CREDENTIALS_FILE="$AWS_CREDENTIALS_FILE" \
+      aws --profile="$AWSPROF" --region="$AWS_REGION" \
+      ssm start-session --target="$i" \
+      --document=AWS-StartInteractiveCommand \
+      --parameters="command='sudo apt-get --yes install netcat'"
+  else
+    AWS_SHARED_CREDENTIALS_FILE="$AWS_CREDENTIALS_FILE" \
+      aws --profile="$AWSPROF" --region="$AWS_REGION" \
+      ssm start-session --target="$i" \
+      --document=AWS-StartInteractiveCommand \
+      --parameters="command='sudo dnf --assumeyes install netcat'"
+  fi
+}
```
If you install the newer `ncat`/`nmap-ncat` then you can use the long-form CLI options.
```suggestion
function installNC() {
  if [[ "$OS" == "Debian" ]]; then
    AWS_SHARED_CREDENTIALS_FILE="$AWS_CREDENTIALS_FILE" \
      aws --profile="$AWSPROF" --region="$AWS_REGION" \
      ssm start-session --target="$i" \
      --document=AWS-StartInteractiveCommand \
      --parameters="command='sudo apt-get --yes install ncat'"
  else
    AWS_SHARED_CREDENTIALS_FILE="$AWS_CREDENTIALS_FILE" \
      aws --profile="$AWSPROF" --region="$AWS_REGION" \
      ssm start-session --target="$i" \
      --document=AWS-StartInteractiveCommand \
      --parameters="command='sudo dnf --assumeyes install nmap-ncat'"
  fi
}
```

bra1ncramp replied:
Are long-form options that critical? I know you like them, but do we have to install a different package just so that we can use a long-form option? I'm confused on the deeper requirement here.

jsf9k replied:
It's just a suggestion.  I was wondering why I had long-form options on my Linux machine and you didn't, then I realized that was because you are installing a different `netcat` package.

bra1ncramp replied:
I would say that unless there are benefits beyond long-form options, I would say use the standard `netcat`.

jsf9k reviewed (commented):
Here is a suggestion.

bra1ncramp committed:
Remove unneeded capitalization

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Improve comment

This may go away - but committing the suggestion for now until we improve the language overall.

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Use `StartNonInteractiveCommand`

Simply running `hostname` should work fine with `AWS-StartNonInteractiveCommand`

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Improve error message

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Fix typo, remove unnecessary capitalization, use long name for flag

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Use correct flag

macOS does not have long-form options for `tee`

bra1ncramp committed:
Fix typo and add more info

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Remove unnecessary capitalization

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Use our standard comment spacing

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Add port number to output message

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Remove an unneeded word from a comment and an output message

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Remove unnecessary capitalization

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Fix typo and remove unnecessary capitalization

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Fix typo

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Use our standard comment spacing

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Improve script usage blurb

bra1ncramp committed:
Use correct environmental variable name

bra1ncramp committed:
Reword and fix typo

bra1ncramp committed:
Use long-form options where available

bra1ncramp committed:
Use long-form options where available

bra1ncramp committed:
Declare environmental variables on command line

Instead of placing AWS environmental variables in the script, declare them on the command line.

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,160 @@
+#!/bin/bash
+#
+# This script will scan AWS instances occurances of IOC hashes listed in the blob at src/ioc_scan/ioc_scanner.py
```
Fix typo and rewrap comment:
```suggestion
# This script will scan AWS instances for occurrences of IOC hashes listed in
# the blob at src/ioc_scan/ioc_scanner.py
```

jsf9k replied:
@bra1ncramp - Where are you testing this?  Can you try changing
```hcl
  # Allow the user to terminate his or her own sessions
  statement {
    actions = [
      "ssm:TerminateSession",
    ]
    resources = [
      "arn:aws:ssm:${var.aws_region}:${local.this_account_id}:session/${aws:username}-*",
    ]
  }
```
to
```hcl
  # Allow the user to terminate his or her own sessions
  statement {
    actions = [
      "ssm:TerminateSession",
    ]
    resources = [
      "arn:aws:ssm:${var.aws_region}:${local.this_account_id}:session/${aws:username}-*",
    ]
  }
```
in the policy `StartStopSSMSession`?  (Note that the only change is that the ampersand (`&`) is changed to a dollar sign (`$`) toward the end of the resource ARN.)  You can simply make this change live, in the AWS console for now.

bra1ncramp committed:
Improve comment and fix spelling

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Check for environmental variable and fix comment

jsf9k committed:
Update shell syntax to satisfy shfmt linter

jsf9k commented:
@bra1ncramp - Are you running `pre-commit` locally?  It should have automatically caught the correction I made in commit 93e1b32.

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,183 @@
+# The filename specified in the first argument
+# (instance-list-file) should contain a list of instance id strings, one per line.
+#
+# The following environmental varialbles must be set:
+# AWS_SHARED_CREDENTAILS_FILE
```
Fix typos and prefer more standard "environment" terminology vs. "environmental":
```suggestion
# The following environment variables must be set:
# AWS_SHARED_CREDENTIALS_FILE
```

dav3r commented on extras/ioc_hash_scan.sh:
```diff
@@ -0,0 +1,183 @@
+  exit 1
+fi
+
+# Check if environmental variables are set
+CRED=$(env | grep AWS_SHARED_CREDENTIALS_FILE)
+REG=$(env | grep AWS_REGION)
+PROF=$(env | grep AWS_PROFILE)
+[[ -z "$CRED" ]] && {
+  echo "AWS_SHARED_CREDENTIALS_FILE environmental variable is not set."
+  exit 1
+}
+[[ -z "$REG" ]] && {
+  echo "AWS_REGION environmental variable is not set."
+  exit 1
+}
+[[ -z "$PROF" ]] && {
+  echo "AWS_PROFILE environmental variable is not set."
```
Prefer more standard "environment" terminology vs. "environmental":
```suggestion
# Check if environment variables are set
CRED=$(env | grep AWS_SHARED_CREDENTIALS_FILE)
REG=$(env | grep AWS_REGION)
PROF=$(env | grep AWS_PROFILE)
[[ -z "$CRED" ]] && {
  echo "AWS_SHARED_CREDENTIALS_FILE environment variable is not set."
  exit 1
}
[[ -z "$REG" ]] && {
  echo "AWS_REGION environment variable is not set."
  exit 1
}
[[ -z "$PROF" ]] && {
  echo "AWS_PROFILE environment variable is not set."
```

dav3r reviewed (commented):
Here are a couple more small thangs to clean up.

bra1ncramp committed:
Fix typos and prefer more standard "environment" terminology vs. "environmental"

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Prefer more standard "environment" terminology vs. "environmental"

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

bra1ncramp committed:
Use terminate-session instead of killing session

bra1ncramp committed:
Do not archive extendend file attributes

bra1ncramp committed:
Don't suppress stderr of pkill command

We've removed the pkill command, so do not need this anymore

jsf9k commented:
> @bra1ncramp - Are you running `pre-commit` locally? It should have automatically caught the correction I made in commit [93e1b32](https://github.com/cisagov/ioc-scanner/commit/93e1b328e0cdbe4d634df79236ba623738ac568c).

@bra1ncramp - Are you running `pre-commit` locally? It should have automatically caught the linting error you committed in e21a303227373d33b7b8b48d6a8d71c335fa419d.

bra1ncramp commented:
> > @bra1ncramp - Are you running `pre-commit` locally? It should have automatically caught the correction I made in commit [93e1b32](https://github.com/cisagov/ioc-scanner/commit/93e1b328e0cdbe4d634df79236ba623738ac568c).
> 
> @bra1ncramp - Are you running `pre-commit` locally? It should have automatically caught the linting error you committed in [e21a303](https://github.com/cisagov/ioc-scanner/commit/e21a303227373d33b7b8b48d6a8d71c335fa419d).

~~What's the error? `pre-commit` isn't showing me any error.~~

I was running it wrong. I should have run `pre-commit run --all-files` instead of just `pre-commit`

bra1ncramp committed:
Remove whitespace for linting

jsf9k commented:
> > > @bra1ncramp - Are you running `pre-commit` locally? It should have automatically caught the correction I made in commit [93e1b32](https://github.com/cisagov/ioc-scanner/commit/93e1b328e0cdbe4d634df79236ba623738ac568c).
> > 
> > 
> > @bra1ncramp - Are you running `pre-commit` locally? It should have automatically caught the linting error you committed in [e21a303](https://github.com/cisagov/ioc-scanner/commit/e21a303227373d33b7b8b48d6a8d71c335fa419d).
> 
> ~What's the error? `pre-commit` isn't showing me any error.~
> 
> I was running it wrong. I should have run `pre-commit run --all-files` instead of just `pre-commit`

You should run `pre-commit install` to install the `git` hook.  Then `pre-commit` will check your changes before you are allowed to commit them.  See the instructions [here](https://github.com/cisagov/ioc-scanner/blob/develop/CONTRIBUTING.md#setting-up-pre-commit), for example.

dav3r reviewed (approved):
Looks solid - strong work! 💪 💼
